implemented perform_y

gsilviHQS opened the pull request:
This is the first step in solving #7  implementing Ry gates (and possibly Rx,Rz)
Also added 2 new colors to handle the complex phase.

gsilviHQS committed:
implemented perform_y
added 2 new colors to handle the complex phase

cduck commented on feynman_path/diagram.py:
```diff
@@ -4,7 +4,7 @@
 from sympy.printing.latex import latex
 import drawSvg as draw
 import latextools
-
```
No need to do this.  Just use pi from `import math`.

cduck commented on feynman_path/diagram.py:
```diff
@@ -175,10 +175,41 @@ def gate_arrow(self, g, time1, key1, key2, amp=1):
-        color = '#26f'
-        if abs(float(amp) - abs(float(amp))) >= 1e-8:
-            color = '#e70'
+        #map the amplitude to  a "wheel" of colors
```
```suggestion
        # map the amplitude to  a "wheel" of colors
```

cduck commented on feynman_path/diagram.py:
```diff
@@ -175,10 +175,41 @@ def gate_arrow(self, g, time1, key1, key2, amp=1):
+        
         self.straight_arrow(g, color, xx1, yy1, xx2, yy2, width=w)
+    
+    def wheel_color(self,amp):
```
Do you think this is enough colors for when you do Ry?  Alternatively, use some continuous color map like HSV where hue is controlled by the phase (i.e. [drawSvg.color.Hsv](https://github.com/cduck/drawSvg/blob/master/drawSvg/color.py))

gsilviHQS replied:
You are right, I was planning to update to a more continuous map later on when implementing the R* gates. I ll check the HSV

cduck commented on feynman_path/diagram.py:
```diff
@@ -258,3 +289,19 @@ def perform_x(self, q_i, *, pre_latex=f'', name='X'):
+                new_state[new_key] = 0
+            new_state[new_key] += new_amp
+        self.transition_text(self.d, t, f'{pre_latex}{name}_{{{q_i}}}')
+        self.add_states(new_state)
```
Nice.  If you intend to add many more gates, can you refactor the perform_* methods to avoid the code copy?

gsilviHQS replied:
Will do. Shall I make a perform_gate class? Or do you have something better in mind?

cduck replied:
It could be as simple as a dictionary with symbolic gate unitaries (at least for single-qubit gates: `{'X': [[0,1],[1,0]], ...}`).  A perform-gate class may just add more boilerplate code.

cduck reviewed (changes requested):
Looks good, only come minor comments.

I will be out until the end of July so unfortunately I won't be able to review more of your code until August.  Feel free to make more PRs for when I'm back.

gsilviHQS committed:
Update feynman_path/diagram.py

Co-authored-by: Casey Duckering <cduck@users.noreply.github.com>

gsilviHQS committed:
factorized single gates + color wheel
also command accept floats
and add simplify to amplitudes.
In case the amplitude is too complicated return the value in digits

gsilviHQS committed:
Merge branch 'support_Ry_gates' of https://github.com/gsilviHQS/feynman_path into support_Ry_gates

gsilviHQS commented:
I made the following changes:
- factorized the single gates to reduce the code replication.
- add a continuous color wheel for amplitudes
- add Rx,Ry,Rz gate with half_turns
- the command from terminal can accept floats
- add a simplify to the amplitudes to avoid labels too complex ( this part add some overhead but I didn't find a better way to deal with it atm), and in case the simplification doesn't reduce enough it translates the symbolic amplitudes to simple digits (too keep a short label)
- fixed a couple of mistakes in the readme examples

gsilviHQS committed:
better visualizatio in my opinion.
where you can follow the sign through the color of the lines

gsilviHQS committed:
color always bright
